lib/dbstructure.rb

```diff
@@ -79,51 +82,188 @@ def create_db_structure(db)
 end
 
 class User
-  attr_accessor :id
-  attr_accessor :userid
-  attr_accessor :username
-  attr_accessor :privlevel
-  def initialize(id, userid, username, privlevel)
-    @id = id
-    @userid = userid
-    @username = username
-    @privlevel = privlevel
+  attr_reader :id
+  attr_reader :userid
+  attr_reader :username
+  attr_reader :privlevel
+
+  def initialize(dbl, userid, privlevel = nil, username = nil)
+    if (not (user = dbl.get_user_by_id(userid)).nil?)
+      @dbl = dbl
+      @id = user[0]
+      @userid = user[1]
+      @username = user[2]
+      @privlevel = user[3]
+      return
+    end
+    if (username != nil and privlevel != nil)
+      @dbl = dbl
+      @id = dbl.add_user(userid, privlevel, username)
+      @userid = userid
+      @username = username
+      @privlevel = privlevel
+      return
+    end
+    @id = nil
   end
+
+  def update_name(name)
+    @dbl.update_user(@userid, @privlevel, name)
+    @name = name
+  end
+  
+  def nth_question(examid, n) = @dbl.user_nth_question(examid, @userid, n)
+  def n_reviews = @dbl.nreviews @userid
+  def all_answers = @dbl.user_all_answers @id
+  def to_userquestion(revid) = @dbl.urid_to_uqid(@userid, revid)
+  def is_priviledged = (@privlevel == 0) ? true : false
+  def is_valid = (@id.nil?) ? false : true
 end
 
 class Question
-  attr_accessor :id
-  attr_accessor :number
-  attr_accessor :variant
-  attr_accessor :text
-  def initialize(id, number, variant, text)
-    @id = id
-    @number = number
-    @variant = variant
-    @text = text
+  attr_reader :id
+  attr_reader :number
+  attr_reader :variant
+  attr_reader :text
+
+  def initialize(dbl, number, variant, text = nil)
+    if (not (question = dbl.get_question(number, variant)).nil?)
+      @dbl = dbl
+      @id = question[0]
+      @number = question[1]
+      @variant = question[2]
+      @text = question[3]
+      return 
+    end
+    if (text != nil)
+      @dbl = dbl
+      @id = dbl.add_question(number, variant, text)
+      @number = number
+      @variant = variant
+      @text = text
+      return
+    end    
+    raise DBLayerError.new "tried to get unregistered question"
+  end
+end
+
+class Exam
```

Comment:
I think to_s method in all such new classes will be good for logging.

Reply:
addressed